emulator/src/kreator/preprocessor.rs

```diff
@@ -180,18 +189,167 @@ fn replace_macros(code: &Vec<String>) -> Result<Vec<String>, &'static str> {
+
+        // check if current block of code is (not) a macro
+        if owned_line.eq(MACRO_START) {
+            in_macro = true;
+            continue;
+        } 
+        if owned_line.eq(MACRO_END) {
+            in_macro = false;
+            label_map.clear();
+            equ_map.clear();
+            set_map.clear();
+            continue;
+        }
+
+        // find set assignments outside of macros
+        if owned_line.contains(" SET ") && !in_macro {
+            let (name, _) = owned_line.split_once(" SET ").unwrap();
+            if !var_name_regex.is_match(line) {
+                return Err("Illegal variable name!");
+            }
+            if !found_set_names.contains(&name.to_string()) {
+                found_set_names.push(name.to_string());
+                all_existing_names.push(name.to_string());
+                label_count += 1;
+            }
+        }
+
+        // change line of macro according to principles of locality
+        if in_macro {
+            
+            // map local labels in macros
+            if loc_label_regex.is_match(&owned_line) && !glob_label_regex.is_match(&owned_line) {
+                let (label, _) = owned_line.split_once(":").unwrap();
+                let (gen_name, generations) = generate_label_name(&all_existing_names, &label_count).unwrap();
+                label_count += generations as u32;
+                all_existing_names.push(gen_name.clone());
+                label_map.insert(label.to_string(), gen_name);
+            }
+
+            // convert globally declared label to normal label
+            if glob_label_regex.is_match(&owned_line) {
+                owned_line = owned_line.replace("::", ":");
+            }
+
+            // map local equ assignments
+            if line.contains(" EQU ") {
+                let (name, _) = owned_line.split_once(" EQU ").unwrap();
+                let (gen_name, generations) = generate_label_name(&all_existing_names, &label_count).unwrap();
+                label_count += generations as u32;
+                all_existing_names.push(gen_name.clone());
+                equ_map.insert(name.to_string(), gen_name);
+            }
+
+            // replace local label calls and equ assigned variables
+            for (old_label, generated_label) in &label_map {
+                owned_line = owned_line.replace(old_label, generated_label);
+            }
+            for (old_name, new_name) in &equ_map {
+                owned_line = owned_line.replace(old_name, new_name);
+            }
+            
+            // check if set assignment variable existed outside of macro
+            if owned_line.contains(" SET ") {
+                let (name, _) = owned_line.split_once(" SET ").unwrap();
+                if !found_set_names.contains(&name.to_string()) {
+                    let (gen_name, generations) = generate_label_name(&all_existing_names, &label_count).unwrap();
+                    label_count += generations as u32;
```

Comment:
Wouldn't the `generations` labels that collided already be included in `label_count`?

Reply:
Could you provide an example? I'm not too sure if I understand your exact point.

Reply:
I think I'm still not quite sure how this needs to work... let's see:

In my mind, to calculate the correct "next" label name we need to do the following:

Given `generated_labels` as the amount of labels generated with our function:
`label_name = (generated_labels/10000 + 'A') + (generated_labels % 10000)`
Then we need to check if `label_name` is containted in `taken_names`, if it is not, we're golden. If it is, we need to:
Increment `generated_labels`, repeat the previous steps.

This seems sufficient to me, I might be missing something. Based on this, I'm having trouble understanding the purpose for this label_count, the different places where it is modified and the reason for subtracting it from the number part of the label.

Reply:
I believe you are perfectly correct with your train of thought. I have implemented it to match your suggestion, simplifying the entire thing drastically. Please see the latest changes to determine if it is what you had in mind.